app/src/main/java/com/github/se/studybuddies/database/DatabaseConnection.kt

```diff
@@ -1293,25 +1282,37 @@ class DatabaseConnection : DbRepository {
     }
   }
 
-  override fun getTimerUpdates(groupUID: String, _timerValue: MutableStateFlow<Long>): Boolean {
-    var isRunning = false
-    groupUID?.let { uid ->
-      val timerRef = rtDb.getReference("timer/$uid")
-      timerRef.addValueEventListener(
-          object : ValueEventListener {
-            override fun onDataChange(snapshot: DataSnapshot) {
-              snapshot.getValue(TimerState::class.java)?.let { timerState ->
-                _timerValue.value = timerState.endTime - System.currentTimeMillis()
-                isRunning = timerState.isRunning
-              }
-            }
+  override fun subscribeToGroupTimerUpdates(
+      groupUID: String,
+      _timerValue: MutableStateFlow<Long>,
+      _isRunning: MutableStateFlow<Boolean>,
+      ioDispatcher: CoroutineDispatcher,
+      mainDispatcher: CoroutineDispatcher,
+      onTimerStateChanged: KSuspendFunction1<TimerState, Unit>
+  ) {
+    groupDataCollection.document(groupUID).addSnapshotListener { snapshot, e ->
+      if (e != null) {
+        Log.w("DatabaseConnection", "Listen failed.", e)
+        return@addSnapshotListener
+      }
 
-            override fun onCancelled(error: DatabaseError) {
-              Log.e("TimerViewModel", "Failed to read timer", error.toException())
-            }
-          })
-    } ?: error("Group UID is not set. Call setup() with valid Group UID.")
-    return isRunning
+      if (snapshot != null && snapshot.exists()) {
+        val timerStateMap = snapshot.get("timerState") as? Map<String, Any>
+        val endTime = timerStateMap?.get("endTime") as? Long ?: 0L
+        val isRunning = timerStateMap?.get("running") as? Boolean ?: false
+        val timerState = TimerState(endTime, isRunning)
+        CoroutineScope(ioDispatcher).launch {
+          val remainingTime = endTime
+          withContext(mainDispatcher) {
+            _timerValue.value = remainingTime
```

Comment:
You could inlined this, remainingTime is set to endTime and then _timerValue.value use it, but only there. So you could remove it and directly set `_timerValue.value = endTime`